MMM Lift test calibration

wd60622 opened the pull request:
<!-- !! Thank your for opening a PR !! -->

<!--- Provide a self-contained summary of your changes in the Title above -->
<!--- This is what will be shown in the automatic release notes: https://github.com/pymc-labs/pymc-marketing/releases -->

## Description
<!--- Describe your changes in detail -->

## Related Issue
<!--- It is good practice to first open an issue explaining the bug / new feature that is addressed by this PR -->
<!--- Please type an `x` in one of the boxes below and provide the issue number after the # sign: -->
- [ ] Closes #
- [ ] Related to #

## Checklist
<!--- Make sure you have completed the following steps before submitting your PR -->
<!--- Feel free to type an `x` in all the boxes below to let us know you have completed the steps: -->
- [ ] Checked that [the pre-commit linting/style checks pass](https://docs.pymc.io/en/latest/contributing/python_style.html)
- [ ] Included tests that prove the fix is effective or that the new feature works
- [ ] Added necessary documentation (docstrings and/or example notebooks)
- [ ] If you are a pro: each commit corresponds to a [relevant logical change](https://wiki.openstack.org/wiki/GitCommitMessages#Structural_split_of_changes)
<!--- You may find this guide helpful: https://mainmatter.com/blog/2021/05/26/keeping-a-clean-git-history/ -->

## Modules affected
<!--- Please list the modules that are affected by this PR by typing an `x` in the boxes below: -->
- [ ] MMM
- [ ] CLV
<!--- Additionally, if you are a maintainer or reviewer, please make sure that the appropriate labels are added to this PR -->

## Type of change
<!--- Select one of the categories below by typing an `x` in the box -->
- [ ] New feature / enhancement
- [ ] Bug fix
- [ ] Documentation
- [ ] Maintenance
- [ ] Other (please specify):
<!--- Additionally, if you are a maintainer or reviewer, please make sure that the appropriate labels are added to this PR -->

wd60622 commented on pymc_marketing/mmm/lift_test.py:
```diff
@@ -0,0 +1,330 @@
+        x_before, x_after, partial_saturation_function
+    )
+
+    dist(
+        name=name,
+        mu=empirical_lift,
+        sigma=df_lift_test["confidence"].to_numpy(),
+        observed=df_lift_test["delta_y"].to_numpy(),
+    )
```
Really only supports distributions with mu and sigma kwargs

wd60622 reviewed (commented):
Some context

Planning on:

- adding to the docstrings
- add lift test module to docs
- example in notebook

wd60622 commented:
Here's an example graph with lift tests based on the DelayedSaturatedMMM class

![model-graph](https://github.com/pymc-labs/pymc-marketing/assets/57733339/ac148a03-c554-4b78-96b7-26df5ccf332e)

cetagostini-wise commented on pymc_marketing/mmm/delayed_saturated_mmm.py:
```diff
@@ -1146,3 +1151,100 @@ def sample_posterior_predictive(
+                "x": [1, 1],
+                "delta_x": [0.1, 0.2],
+                "delta_y": [0.1, 0.1],
+                "confidence": [0.1, 0.1],
```
A few side notes:

1. Confidence should be the `sigma` define for an experiment, no?
2. Tools like Causal Impact (Example) provide the sigma around the mean estimate, and ideally, this value should be equal to the standard deviation given by the experiment framework, right?
3. What criteria are used to define confidence? In case a tool like `causalpy` doesn't deliver it to you, what are possible frameworks or recommended values?

It's important to clarify the definition of confidence and demonstrate how having different degrees of reliability can impact the results of an experiment. I think that marketers are not very familiar with calibration and those questions may arise!

wd60622 replied:
Thanks for taking a look @carlosagostini 

1. Yes, correct. I changed the name to `sigma` from `confidence`. I think that would be more clear
2. Do you have a link for this example?
3. Good point to point out any conversions needed or 

I have started a notebook

cetagostini-wise replied:
Causal impact example output:

```python
## Posterior inference {CausalImpact}
## 
##                          Average        Cumulative  
## Actual                   117            3511        
## Prediction (s.d.)        107 (0.37)     3196 (11.03)
## 95% CI                   [106, 107]     [3174, 3217]
##                                                     
## Absolute effect (s.d.)   11 (0.37)      316 (11.03) 
## 95% CI                   [9.8, 11]      [294.9, 337]
##                                                     
## Relative effect (s.d.)   9.9% (0.35%)   9.9% (0.35%)
## 95% CI                   [9.2%, 11%]    [9.2%, 11%] 
## 
## Posterior tail-area probability p:   0.001
## Posterior prob. of a causal effect:  99.9%
```

juanitorduz commented on pymc_marketing/mmm/delayed_saturated_mmm.py:
```diff
@@ -1146,3 +1150,87 @@ def sample_posterior_predictive(
+                "x": [1, 1],
+                "delta_x": [0.1, 0.2],
+                "delta_y": [0.1, 0.1],
+                "confidence": [0.1, 0.1],
```
Shall we change the example docstring from confidence -> sigma?

juanitorduz commented on pymc_marketing/mmm/delayed_saturated_mmm.py:
```diff
@@ -1146,3 +1150,87 @@ def sample_posterior_predictive(
+        df_lift_test_scaled = scale_lift_measurements(
+            df_lift_test=df_lift_test,
+            channel_col="channel",
+            channel_columns=self.channel_columns,  # type: ignore
```
why the type ignore?

wd60622 replied:
something downstream of type like list[str]

juanitorduz commented on pymc_marketing/mmm/lift_test.py:
```diff
@@ -0,0 +1,366 @@
+
+
+def _lift_test_index(
+    lift_values: np.ndarray, model_values: np.ndarray
+) -> npt.NDArray[np.int_]:
```
I think we should keep the same type hit convention for np arrays `np.ndarray` or `npt.NDArray`. Mixing them look confusing ;)

juanitorduz commented on pymc_marketing/mmm/lift_test.py:
```diff
@@ -0,0 +1,366 @@
+    return np.argmax(same_value, axis=1)
+
+
+Index = npt.NDArray[np.int_]
```
Is the return type of ` _lift_test_index` `Index`?

wd60622 replied:
good catch

juanitorduz commented on pymc_marketing/mmm/lift_test.py:
```diff
@@ -0,0 +1,366 @@
+    Assumes any column in the DataFrame is a coordinate in the model with the
+    same name.
+
+    Args:
+        df_lift_test: DataFrame with lift test results.
+        model: PyMC model with date and channel coordinates.
+
+    Returns:
+        Dictionary of indices for the lift test results in the model.
```
I think we are following the numpy docstring format right? https://numpydoc.readthedocs.io/en/latest/format.html

juanitorduz commented on pymc_marketing/mmm/lift_test.py:
```diff
@@ -0,0 +1,366 @@
+def empirical_lift_measurements(x_before, x_after, saturation_curve, pt=pt):
+    """Calculate the empirical lift measurements at two spends.
+
+    Args:
```
I think we are following the numpy docstring format right? https://numpydoc.readthedocs.io/en/latest/format.html

juanitorduz commented on pymc_marketing/mmm/lift_test.py:
```diff
@@ -0,0 +1,366 @@
+    }
+
+
+def empirical_lift_measurements(x_before, x_after, saturation_curve, pt=pt):
```
type hints?

juanitorduz commented on pymc_marketing/mmm/lift_test.py:
```diff
@@ -0,0 +1,366 @@
+def required_dims_from_named_vars_to_dims(
+    named_vars_to_dims: dict[str, tuple[str, ...]],
+) -> list[str]:
+    """Get the required dimensions from a named_vars_to_dims dictionary.
```
I think we are following the numpy docstring format right? https://numpydoc.readthedocs.io/en/latest/format.html

juanitorduz commented on pymc_marketing/mmm/lift_test.py:
```diff
@@ -0,0 +1,366 @@
+) -> Indices:
+    """Get the indices of the lift test results in the model.
+
+    Args:
```
I think we are following the numpy docstring format right? https://numpydoc.readthedocs.io/en/latest/format.html

juanitorduz commented on pymc_marketing/mmm/lift_test.py:
```diff
@@ -0,0 +1,366 @@
+    model: pm.Model,
+    var_names: list[str],
+) -> Indices:
+    """Get the indices of the lift test results in the model.
```
Could you explain better what an  "index" is?

wd60622 replied:
added below. Let me know what you think
I can rename the function if you'd like

juanitorduz commented on pymc_marketing/mmm/lift_test.py:
```diff
@@ -0,0 +1,366 @@
+    """Get the indices of the lift test results in the model.
+
+    Args:
+        df_lift_test: DataFrame with lift test results.
```
It would be good to provide the desired structure of the dataframe: size? columns? types?

juanitorduz commented on pymc_marketing/mmm/lift_test.py:
```diff
@@ -0,0 +1,366 @@
+) -> None:
+    """Add lift measurements to the likelihood of the model.
+
+    Args:
```
I think we are following the numpy docstring format right? https://numpydoc.readthedocs.io/en/latest/format.html

juanitorduz commented on pymc_marketing/mmm/lift_test.py:
```diff
@@ -0,0 +1,366 @@
+    """Add lift measurements to the likelihood of the model.
+
+    Args:
+        df_lift_test: DataFrame with lift test results.
```
Same comment as above regarding the expected structure of the dataframe.

juanitorduz commented on pymc_marketing/mmm/lift_test.py:
```diff
@@ -0,0 +1,366 @@
+) -> None:
+    """Add empirical lift measurements to the likelihood of the model.
+
+    Args:
```
I think we are following the numpy docstring format right? https://numpydoc.readthedocs.io/en/latest/format.html

juanitorduz commented on pymc_marketing/mmm/lift_test.py:
```diff
@@ -0,0 +1,366 @@
+    """Add empirical lift measurements to the likelihood of the model.
+
+    Args:
+        df_lift_test: DataFrame with lift test results.
```
Same comment about df structure 😅

juanitorduz commented on pymc_marketing/mmm/lift_test.py:
```diff
@@ -0,0 +1,366 @@
+        model: PyMC model with date and channel coordinates.
+        name: Name of the likelihood.
+
+    Returns:
+        None
```
I think we can remove this

juanitorduz commented on pymc_marketing/mmm/lift_test.py:
```diff
@@ -0,0 +1,366 @@
+) -> None:
+    """Add empirical lift measurements to the likelihood of the model.
+
+    Args:
```
I think we are following the numpy docstring format right? https://numpydoc.readthedocs.io/en/latest/format.html

juanitorduz commented on pymc_marketing/mmm/lift_test.py:
```diff
@@ -0,0 +1,366 @@
+    """Add empirical lift measurements to the likelihood of the model.
+
+    Args:
+        df_lift_test: DataFrame with lift test results.
```
Structure of the data frame in docstrings.

juanitorduz commented on pymc_marketing/mmm/lift_test.py:
```diff
@@ -0,0 +1,366 @@
+) -> pd.DataFrame:
+    """Scale the lift measurements for a specific channel.
+
+    Args:
```
I think we are following the numpy docstring format right? https://numpydoc.readthedocs.io/en/latest/format.html

juanitorduz commented on pymc_marketing/mmm/lift_test.py:
```diff
@@ -0,0 +1,366 @@
+    """Scale the lift measurements for a specific channel.
+
+    Args:
+        df_lift_test: DataFrame with lift test results.
```
Structure df description.

juanitorduz commented on pymc_marketing/mmm/lift_test.py:
```diff
@@ -0,0 +1,366 @@
+        channel_col, append=True
+    )
+    df_to_rescale = (
+        df_original.stack()
+        .unstack(channel_col)
```
why stack and unstack? 🤔

wd60622 replied:
This DataFrame has the MultiIndex with two levels. (RangeIndex and channel_column)
After the stack, there is 3 levels to the MultiIndex. (RangeIndex, channel_column, (x, delta_x))
The unstack makes the channel_column the columns rather than the x, delta_x

All this to make use of the scikit-learn transformers with the columns it is trained on

juanitorduz replied:
Thanks!

juanitorduz replied:
Maybe add it as a small inline comment "there is 3 levels to the MultiIndex. (RangeIndex, channel_column, (x, delta_x))" for future Will and Juan (an others)

juanitorduz commented on pymc_marketing/mmm/lift_test.py:
```diff
@@ -0,0 +1,366 @@
+    df_to_rescale = (
+        df_original.stack()
+        .unstack(channel_col)
+        .reindex(channel_columns, axis=1)  # type: ignore
```
why type hint ignore?

wd60622 replied:
It thought it was a Series but it is a DataFrame

wd60622 replied:
I pulled it out into a helper function that is used twice. Also put some comments for future devs

juanitorduz commented on pymc_marketing/mmm/lift_test.py:
```diff
@@ -0,0 +1,366 @@
+) -> pd.Series:
+    """Scale the target for the lift measurements.
+
+    Args:
```
I think we are following the numpy docstring format right? https://numpydoc.readthedocs.io/en/latest/format.html

juanitorduz reviewed (changes requested):
@wd60622 As always reviewing your code is a pleasure! I love the modular code and the descriptive names. My comments are more about docstrings and style so that is consistent with the whole code base.

I will now take a look into the notebook :)

juanitorduz commented on docs/source/notebooks/mmm/mmm_lift_test.ipynb:
```diff
@@ -0,0 +1,1716 @@
+{
```
<p>Can you please add <code>%config InlineBackend.figure_format = "retina"</code> to make the plots of higher resolution 🤗 </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pymc-labs/pymc-marketing/pull/590/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='2'/>

wd60622 replied:
<p>done!</p>

juanitorduz commented on docs/source/notebooks/mmm/mmm_lift_test.ipynb:
```diff
@@ -0,0 +1,1716 @@
+{
```
<p>Add something like: We use the <code>DelayedSaturatedMMM</code> class to specify our model the usual way (or something similar) </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pymc-labs/pymc-marketing/pull/590/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='9'/>

wd60622 replied:
<p>done</p>

juanitorduz commented on docs/source/notebooks/mmm/mmm_lift_test.ipynb:
```diff
@@ -0,0 +1,1716 @@
+{
```
<p>Can you explain that yoiu are generating synthetic data with some true regression coefficients? So that the user follows the storyline of the code :)</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pymc-labs/pymc-marketing/pull/590/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='10'/>

wd60622 replied:
<p>good point</p>

juanitorduz commented on docs/source/notebooks/mmm/mmm_lift_test.ipynb:
```diff
@@ -0,0 +1,1716 @@
+{
```
<p>Can you put the legend outside the plot (<a href="https://stackoverflow.com/questions/4700614/how-to-put-the-legend-outside-the-plot" target="_blank">https://stackoverflow.com/questions/4700614/how-to-put-the-legend-outside-the-plot</a>) and add a title?</p><p> </p><p>Also what does "column value" mean (maybe use the same terminology as the mmm example notebook)?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pymc-labs/pymc-marketing/pull/590/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='13'/>

wd60622 replied:
<p>Removed xlabel and added a descriptive title</p>

juanitorduz commented on docs/source/notebooks/mmm/mmm_lift_test.ipynb:
```diff
@@ -0,0 +1,1716 @@
+{
```
<p>You coult try adding the true values as dshed vertical lines as in the mmm example notebook</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pymc-labs/pymc-marketing/pull/590/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='18'/>

juanitorduz replied:
<p>Maybe you comment these are auxiliary plotting functions which are not that important. You can even add the tag <code>hide-input</code> to this cell to hide the code </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pymc-labs/pymc-marketing/pull/590/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='27'/>

juanitorduz replied:
<p>This is absolutely fantastic Will!</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pymc-labs/pymc-marketing/pull/590/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='32'/>

juanitorduz replied:
<p>Can you please add some final words and main takeaways :)</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pymc-labs/pymc-marketing/pull/590/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='49'/>

juanitorduz commented on docs/source/notebooks/mmm/mmm_lift_test.ipynb:
```diff
@@ -0,0 +1,1716 @@
+{
```
<p>Can you explain a bit more the how to read these triangles? </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pymc-labs/pymc-marketing/pull/590/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='28'/>

wd60622 replied:
<p>Added some more in the markdown above it</p>

juanitorduz commented on docs/source/notebooks/mmm/mmm_lift_test.ipynb:
```diff
@@ -0,0 +1,1716 @@
+{
```
<p>We can also hide the code of these functions</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pymc-labs/pymc-marketing/pull/590/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='39'/>

wd60622 replied:
<p>done</p>

juanitorduz commented on docs/source/notebooks/mmm/mmm_lift_test.ipynb:
```diff
@@ -0,0 +1,1716 @@
+{
```
<p>Missing "." end of the sentence ;)</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pymc-labs/pymc-marketing/pull/590/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='42'/>

wd60622 replied:
<p>done. Thank!</p>

juanitorduz commented:
@wd60622 I love your notebook! I suggest we add more story to it. Otherwise looks amazing!

juanitorduz commented on docs/source/notebooks/mmm/mmm_lift_test.ipynb:
```diff
@@ -0,0 +1,1716 @@
+{
```
<p>Could you also include the outline of the notebook? and maybe an example of what a lift test is (maybe say something about CausalPy?)</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pymc-labs/pymc-marketing/pull/590/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='0'/>

wd60622 replied:
<p>There is an outline that will be autogenerated in the documentation based on the headers in the notebook. Do you think more is required?</p>

twiecki commented on docs/source/notebooks/mmm/mmm_lift_test.ipynb:
```diff
@@ -0,0 +1,1716 @@
+{
```
<p>Lift tests will be used to alter our model likelihood and  ... ?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pymc-labs/pymc-marketing/pull/590/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='24'/>

wd60622 replied:
<p>good catch. Must of got distracted</p>

cetagostini commented:
@wd60622 amazing work, I have a few quick recommendations:

1. What if we add an introduction? E.g:
```markdown
# Introduction
You may have heard of the phrase "all models are wrong; some models are useful." This is true in many areas, and it's likely that after the first attempt, you haven't yet created a model that can accurately determine your true attribution. Even if you have, how can you be sure? We create models to understand the attribution of our marketing channels, but it appears that even then, we can't always rely on what the models tell us.

In order to ensure that our models are decomposing correctly, we can use various testing methods to gather real-world data and compare it with our models. This will help us to identify any discrepancies and improve the decomposition accuracy of our models. 

Today, we will explore a new way to integrate experiments into **pymc-marketing**. This will bring us closer to accurate representations of real-world values and improve the estimates generated by our model.

# Requirements
Today, we won't be discussing how to conduct various tests, but instead, we will focus on their utilization. If you wish to acquire knowledge on how to generate results that are compatible with your MMM models, you can check out CausalPy by clicking on this link.

# Goal
After reading this article, you will have gained the necessary expertise to incorporate the results (detected uplift from your experiments) into our regressive model.
```

2. What if we add some reflections? e.g:
- The current version of `pymc-marketing` lacks the capability to incorporate time-varying coefficients. Attempting to adjust likelihood with a single experiment can be counterproductive as it may inform the average performance of a channel based on a specific result at a particular moment in time, extracted from the experiment. To avoid any strange biases, we recommend conducting at least 2 or 3 experiments for the calibration process.

3. What if we add some final conclusion? As @juanitorduz mentioned.

I really enjoyed reading the notebook, but I think we could improve it by adding more storytelling. Although all people here can understand it, others who are new to this topic or have no prior knowledge might find it difficult to grasp the idea. To make it more accessible to everyone, we could add examples or metaphors to explain it less technically.

Amazing contribution! 🔥

wd60622 commented on pymc_marketing/mmm/delayed_saturated_mmm.py:
```diff
@@ -1146,3 +1150,87 @@ def sample_posterior_predictive(
+    def add_lift_test_measurements(
+        self,
+        df_lift_test: pd.DataFrame,
+        dist=pm.Gamma,
```
I am going with `pm.Distribution`. Not sure if that is the best type hint though

wd60622 commented on pymc_marketing/mmm/delayed_saturated_mmm.py:
```diff
@@ -1146,3 +1150,87 @@ def sample_posterior_predictive(
+        dist=pm.Gamma,
+        name: str = "lift_measurements",
+    ) -> None:
+        """Add lift tests to the model.
```
Added a text description and some sudo code. Let me know what you think

wd60622 commented:
Should have addressed all of the comments on the code. Please tag me again or reraise if I didn't!
Thanks again for you feedback 🙏 

On to the notebook feedback 💪

wd60622 commented:
Alright @juanitorduz @twiecki @cetagostini 
Thank you all for your eyes!

I have went through and incorporated your wonderful feedback. Thanks!
Please let me know if I have missed anything or if the changes need to be updated more. 

Feel like we are almost there 🚀

juanitorduz commented on pymc_marketing/mmm/lift_test.py:
```diff
@@ -0,0 +1,533 @@
+    return np.argmax(same_value, axis=1)
+
+
+Indices = dict[str, Index]
```
maybe is better to have all the types together? so we move this up?

wd60622 replied:
Done!

juanitorduz reviewed (changes requested):
Small suggestion :) Otherwise LGTM

juanitorduz commented on docs/source/notebooks/mmm/mmm_lift_test.ipynb:
```diff
@@ -0,0 +1,1776 @@
+{
```
<p>Nice use of <code>pm.do</code>!</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pymc-labs/pymc-marketing/pull/590/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='11'/>

juanitorduz commented:
@wd60622 I really like the new version of the notebook! Let me know what you think about my small suggestion above and see if we can merge this one :)

wd60622 commented:
> @wd60622 I really like the new version of the notebook! Let me know what you think about my small suggestion above and see if we can merge this one :)

Thanks for the feedback! @juanitorduz Any remaining items that I missed? 

Are there any issue to close or discussions to point to? I see [this issue](https://github.com/pymc-labs/pymc-marketing/issues/60)
and [this discussion](https://github.com/pymc-labs/pymc-marketing/discussions/488) that I've already commented on

juanitorduz reviewed (approved):
Amazing! 🚀 

Maybe @cetagostini @ulfaslak wanna give a last look ?

ulfaslak commented on docs/source/notebooks/mmm/mmm_lift_test.ipynb:
```diff
@@ -0,0 +1,1776 @@
+    "2. Add lift measurements: `mmm.add_lift_test_measurements(df_lift_test)`\n",
+    "2. Sample posterior: `mmm.fit(X, y)`\n",
+    "\n",
+    "This in a case study of two correlated channels and see how lift tests help distinguish the channel effects."
```
Typos.

Replace:

"This in a case study of two correlated channels and see how lift tests help distinguish the channel effects."

With:

"This is a case study of two correlated channels to see how lift tests help distinguish channel effects."

ulfaslak commented on docs/source/notebooks/mmm/mmm_lift_test.ipynb:
```diff
@@ -0,0 +1,1776 @@
+   "outputs": [
+    {
+     "data": {
+      "image/png": "iVBORw0KGgoAAAANSUhEUgAABLsAAALHCAYAAABmNkB7AAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjguMywgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy/H5lhTAAAACXBIWXMAAA9hAAAPYQGoP6dpAAEAAElEQVR4nOzdd2DTdf748WeS7p1uKFtoGcqWISiKiKio4DqOE+TUcx2cevpz+/X0HKjnOfDOcW7cKA4UFQERUFkFRKTsUdrSvWfS5PP7I/l82tIWkmY2fT3+UT755PN559M0TV55DZ2iKApCCCGEEEIIIYQQQgQAva8XIIQQQgghhBBCCCGEu0iwSwghhBBCCCGEEEIEDAl2CSGEEEI ...
```
Typo:

"usual channel spend"

wd60622 replied:
I'm not sure where this is

ulfaslak commented on docs/source/notebooks/mmm/mmm_lift_test.ipynb:
```diff
@@ -0,0 +1,1776 @@
+   "outputs": [
+    {
+     "data": {
+      "image/png": "iVBORw0KGgoAAAANSUhEUgAABLsAAALHCAYAAABmNkB7AAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjguMywgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy/H5lhTAAAACXBIWXMAAA9hAAAPYQGoP6dpAAEAAElEQVR4nOzdd2DTdf748WeS7p1uKFtoGcqWISiKiKio4DqOE+TUcx2cevpz+/X0HKjnOfDOcW7cKA4UFQERUFkFRKTsUdrSvWfS5PP7I/l82tIWkmY2fT3+UT755PN559M0TV55DZ2iKApCCCGEEEIIIYQQQgQAva8XIIQQQgghhBBCCCGEu0iwSwghhBBCCCGEEEIEDAl2CSGEEEI ...
```
I would be slightly more pedagogical here. I suggest:

## About Lift Tests

In a lift study, one temporarily changes the budget of a channel for a fixed period of time, and then uses some method (for example [CausalPy](https://causalpy.readthedocs.io/en/latest/)) to make inference about the change in sales directly caused by the adjustment.

A lift test is characterized by:

- `channel`: the channel that was tested
- `x`: pre-test channel spend
- `delta_x`: change made to `x` 
- `delta_y`: inferred change in sales due to `delta_x`
- `sigma`: standard deviation of `delta_y`

An experiment characterized in this way can be viewed as two points on the saturation curve for the channel. Accordingly, lift test calibration is implemented by adding a term to the model likelihood, that makes the channel saturation curve (the contribution as a function of spend) align with the test observation.

ulfaslak replied:
Unclear what you mean by this. Consider rephrasing. E.g.

"Having created a `DelayedSaturatedMMM` model instance, `mmm`, and built it using `mmm.build_model(X, y)`/`mmm.fit(X,y)`, we can add a lift test result to the model using the `add_lift_test_measurements` method, directly on the model object."

cetagostini-wise commented:
I think would be valuable the plot where you show how the contribution change with the lift test contribution estimation. Basically, the same you show to us, do you think it is easy to add?

ulfaslak commented on pymc_marketing/mmm/delayed_saturated_mmm.py:
```diff
@@ -1146,3 +1150,112 @@ def sample_posterior_predictive(
+            target_transform=self.target_transformer.transform,
+        )
+        with self.model:
+            add_logistic_empirical_lift_measurements_to_likelihood(
```
Using the term "empirical" correctly here, but...

ulfaslak commented on pymc_marketing/mmm/delayed_saturated_mmm.py:
```diff
@@ -1146,3 +1150,112 @@ def sample_posterior_predictive(
+
+        The empirical difference of a channel's saturation curve is created
+        from `x` and `x + delta_x` for each channel. This random variable is
+        then then conditioned using the `delta_y` and `sigma` of the lift test
```
then then

ulfaslak commented on pymc_marketing/mmm/delayed_saturated_mmm.py:
```diff
@@ -1146,3 +1150,112 @@ def sample_posterior_predictive(
+    ) -> None:
+        """Add lift tests to the model.
+
+        The empirical difference of a channel's saturation curve is created
```
Total nitpicker here, but I think you're using the term "empirical" wrong starting here. What you're actually doing below (line 277) is estimating the "theoretical" or "model estimated" lift. The *empirical* lift if actually `delta_y`.

wd60622 replied:
Thanks. Good feedback. 

I will go with the distinction
1. that the model evaluated difference will be "model_estimated"
2. delta_y = "empirical_lift"

ulfaslak commented on pymc_marketing/mmm/lift_test.py:
```diff
@@ -0,0 +1,531 @@
+    saturation_curve: Callable[[npt.NDArray[np.float_]], npt.NDArray[np.float_]],
+    pt=pt,
+) -> npt.NDArray[np.float_]:
+    """Calculate the empirical lift measurements at two spends.
```
Again, the empirical lift was calculated (for example in CausalPy) when the user curated the lift test result. Here, we rather estimate the "theoretical lift given model". I don't know what a better function name would be. Maybe `estimate_modeled_lift`.

ulfaslak commented:
I could go deeper for sure. Ran your notebook and it works as you demonstrate. I probably missed the discussion on this, but what is the reason for not adding support for test dates and other saturation functions? I'm a `tanh` guy, you know.

juanitorduz commented:
Regarding the other saturation functions. I suggest we work iteratively and make sure we create issues to tackle them afterwards :)

ulfaslak commented:
> Regarding the other saturation functions. I suggest we work iteratively and make sure we create issues to tackle them afterwards :)

Makes sense.

wd60622 commented:
Thanks for the feedback @ulfaslak 
Will make some changes tomorrow

wd60622 commented:
> I think would be valuable the plot where you show how the contribution change with the lift test estimation. Basically, the same you show to us, do you think it is easy to add?

Bit unclear on the ask. Is there a missing visual or should the existing visual be presented in a different way?
Add in what way? As a method to the class or to the notebook?

wd60622 commented:
> > Regarding the other saturation functions. I suggest we work iteratively and make sure we create issues to tackle them afterwards :)
> 
> Makes sense.

Working with @carlosagostini on #602 
The two specific function for menten and logistic will likely be removed and would be built out via the class representation of the saturation functions. 
With that, any saturation function that meets our requirements will be able to be used with lift tests
